src/game/client/draw/w3dtanktruckdraw.cpp

```diff
@@ -0,0 +1,651 @@
+            m_treadDebrisRight->Start();
+        }
+    }
+}
+
+void W3DTankTruckDraw::Stop_Move_Debris()
+{
+    if (m_treadDebrisLeft != nullptr) {
+        m_treadDebrisLeft->Stop();
+    }
+
+    if (m_treadDebrisRight != nullptr) {
+        m_treadDebrisRight->Stop();
+    }
+}
+
+void W3DTankTruckDraw::Update_Tread_Positions(float uv_delta)
+{
+    for (int i = 0; i < m_treadCount; i++) {
+        float offset = 0;
+
+        if (m_treads[i].m_type == TREAD_LEFT) {
+            offset = m_treads[i].m_materialSettings.m_customUVOffset.X + uv_delta;
+        } else if (m_treads[i].m_type == TREAD_RIGHT) {
+            offset = m_treads[i].m_materialSettings.m_customUVOffset.X - uv_delta;
+        }
+
+        offset -= GameMath::Floor(offset);
+        m_treads[i].m_materialSettings.m_customUVOffset.Set(offset, 0.0f);
+    }
+}
+
+void W3DTankTruckDraw::Update_Tread_Objects()
+{
+    RenderObjClass *robj = Get_Render_Object();
+
+    for (int i = 0; i < m_treadCount; i++) {
+        Ref_Ptr_Release(m_treads[i].m_robj);
+    }
+
+    m_treadCount = 0;
+
+    if (Get_W3D_Tank_Truck_Draw_Module_Data() != nullptr
+        && Get_W3D_Tank_Truck_Draw_Module_Data()->m_treadAnimationRate != 0.0f && robj != nullptr) {
+        for (int j = 0; j < robj->Get_Num_Sub_Objects() && m_treadCount < MAX_TREADS_PER_TANK; j++) {
+            RenderObjClass *subobj = robj->Get_Sub_Object(j);
+
+            if (subobj != nullptr) {
+                if (subobj->Class_ID() == RenderObjClass::CLASSID_MESH) {
+                    if (subobj->Get_Name() != nullptr) {
+                        const char *str = strchr(subobj->Get_Name(), '.');
+                        if (str != nullptr && *str++ != 0 && !strncasecmp(str, "TREADS", 6)) {
+                            MaterialInfoClass *matinfo = subobj->Get_Material_Info();
+
+                            if (matinfo) {
+                                for (int index = 0; index < matinfo->Vertex_Material_Count(); index++) {
+                                    TextureMapperClass *mapper = matinfo->Get_Vertex_Material(index)->Get_Mapper(0);
+
+                                    if (mapper != nullptr
+                                        && mapper->Mapper_ID() == TextureMapperClass::MAPPER_ID_LINEAR_OFFSET) {
+                                        static_cast<LinearOffsetTextureMapperClass *>(mapper)->Set_UV_Offset_Delta(
+                                            Vector2(0.0f, 0.0f));
+                                        subobj->Add_Ref();
+                                        m_treads[m_treadCount].m_robj = subobj;
+                                        m_treads[m_treadCount].m_type = TREAD_MIDDLE;
+                                        subobj->Set_User_Data(&m_treads[m_treadCount].m_materialSettings, false);
+                                        m_treads[m_treadCount].m_materialSettings.m_customUVOffset = Vector2(0.0f, 0.0f);
+                                        m_treadCount++;

```

Comment:
Call to `Get_Mapper(0)` calls an AddRef on `TextureMapperClass`. Is there a leak here?

Reply:
Should be Peek_Mapper. Fixed.